compiler/rustc_mir_transform/src/const_prop.rs

```diff
@@ -1027,23 +964,17 @@ impl<'tcx> MutVisitor<'tcx> for ConstPropagator<'_, 'tcx> {
         // We remove all Locals which are restricted in propagation to their containing blocks and
         // which were modified in the current block.
         // Take it out of the ecx so we can get a mutable reference to the ecx for `remove_const`.
-        let mut locals = std::mem::take(&mut self.ecx.machine.written_only_inside_own_block_locals);
-        for &local in locals.iter() {
-            Self::remove_const(&mut self.ecx, local);
-        }
-        locals.clear();
-        // Put it back so we reuse the heap of the storage
-        self.ecx.machine.written_only_inside_own_block_locals = locals;
-        if cfg!(debug_assertions) {
-            // Ensure we are correctly erasing locals with the non-debug-assert logic.
-            for local in self.ecx.machine.only_propagate_inside_block_locals.iter() {
-                assert!(
-                    self.get_const(local.into()).is_none()
-                        || self
-                            .layout_of(self.local_decls[local].ty)
-                            .map_or(true, |layout| layout.is_zst())
-                )
+        let can_const_prop = std::mem::take(&mut self.ecx.machine.can_const_prop);
+        for (local, &mode) in can_const_prop.iter_enumerated() {
+            match mode {
+                ConstPropMode::FullConstProp => {}
```

Comment:
It looks to me like the regression is from this loop? At least that's what my profiler indicates. It looks like basically all the cycles in `ConstProp` are driving this iterator and hitting this first case:
```
 16.79 │ d0:┌─→inc    %rdi
       │    │<core::slice::iter::Iter<rustc_mir_transform::const_prop::ConstPropMode> as core::iter::traits::iterator::Iterator>::next:
 16.05 │    │  add    $0x48,%rdx
 16.52 │    │  cmp    %rdi,%rax
       │    │↓ je     12e
       │    │<rustc_middle::mir::Local>::from_usize: 
       │    │} 
       │    │
       │    │///////////////////////////////////////////////////////////////////////////
       │    │// Variables and temps 
       │    │ 
       │    │rustc_index::newtype_index! {
 25.17 │ dc:│  cmp    %r8,%rdi
       │    │↓ je     1b7
       │    │<rustc_mir_transform::const_prop::ConstPropagator as rustc_middle::mir::visit::MutVisitor>::visit_basic_block_data: 
       │    │match mode {
 25.18 │    ├──cmpb   $0x1,(%rcx,%rdi,1)
       │    └──jne    d0
```

Reply:
I will not be working on this, but I suspect the problem centers around there being a _lot_ of `FullConstProp` in here and we could be iterating over less elements somehow.

Reply:
Thanks for the analysis. I'll handle it after hours.